Source/ExecutionEngine/ExecutionEngine.cs

```diff
@@ -1537,34 +1537,54 @@ private ErrorInformation CreateErrorInformation(Counterexample error, VC.VCGen.O
       {
         Debug.Assert(error is AssertCounterexample);
         var assertError = (AssertCounterexample)error;
-        if (assertError.FailingAssert is LoopInitAssertCmd or LoopInvMaintainedAssertCmd)
+        var failingAssert = assertError.FailingAssert;
+        if (failingAssert is LoopInitAssertCmd or LoopInvMaintainedAssertCmd)
         {
-          errorInfo = ErrorInformationFactory.CreateErrorInformation(assertError.FailingAssert.tok,
-            assertError.FailingAssert.Description.FailureDescription,
+          errorInfo = ErrorInformationFactory.CreateErrorInformation(failingAssert.tok,
+            failingAssert.Description.FailureDescription,
             assertError.RequestId, assertError.OriginalRequestId, cause);
-          errorInfo.Kind = assertError.FailingAssert is LoopInitAssertCmd ?
+          errorInfo.Kind = failingAssert is LoopInitAssertCmd ?
             ErrorKind.InvariantEntry : ErrorKind.InvariantMaintainance;
-          if ((assertError.FailingAssert.ErrorData as string) != null)
+          string relatedMessage = null;
+          if (failingAssert.ErrorData is string)
           {
-            errorInfo.AddAuxInfo(assertError.FailingAssert.tok, assertError.FailingAssert.ErrorData as string,
-              "Related message");
+            relatedMessage = failingAssert.ErrorData as string;
+          }
+          else if (failingAssert is LoopInitAssertCmd initCmd)
+          {
+            var desc = initCmd.originalAssert.Description;
+            if (desc.ShortDescription != "assert")
```

Comment:
Why don't you move this check to where the `LoopInitAssertCmd` is created and then immediately replace its description with the description of the original?

Also, checking against a hardcoded string seems frail. Don't you want to replace "assert" with `AssertionDescription.ShortDescription` ?

Reply:
My thinking is that there may wind up being other aspects of the original assertion that it's useful to access in the future.

And, yeah, checking against AssertionDescription.ShortDescription is probably the right thing. I'll change that.

Reply:
But a type test does work.

Reply:
My suggestion was to make AssertionDescription a singleton, but a type test works too.

> My thinking is that there may wind up being other aspects of the original assertion that it's useful to access in the future.

Why not make that change then?

Reply:
I'm not sure I know what you mean by "that change".

Although `AssertionDescription` itself could be a singleton, other subclasses of `ProofObligationDescription` can't be.

Reply:
"That change" would be adding the "originalAssert" field, as opposed to only using the Description field.

Reply:
> > Although AssertionDescription itself could be a singleton
> 
> That would be enough right?
> 
> > , other subclasses of ProofObligationDescription can't be.
> 
> Why not? None of them take parameters or are mutable right?

The subclasses in Boogie don't take parameters, but most of the ones [in Dafny](https://github.com/dafny-lang/dafny/pull/1915) do.

Reply:
> Have you considered moving this check to where the `LoopInitAssertCmd` is created and then immediately replacing its description with the description of the original? That way you avoid having to introduce the `originalAssert` field.

This is unfortunately not directly possible, since the current logic uses _both_ descriptions: the one associated with the `LoopInitAssertCmd` and the one associated with the `AssertCmd` it was derived from.